File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        D={}
        def traverse(i,j,root,D):
            if root==None:
                return
            else:
                if j not in D:
                    D[j]={i:[root.val]}
                else:
                    T=D[j]
                    if i not in T:
                        T[i]=[root.val]
                    else:
                        T[i].append(root.val)
                if root.left:
                    traverse(i+1,j-1,root.left,D)
                if root.right:
                    traverse(i+1,j+1,root.right,D)
        traverse(0,0,root,D)
        A=list(D.keys())
        A.sort()
        Ans=[]
        for i in A:
            T=D[i]
            temp=list(T.keys())
            temp.sort()
            arr=[]
            for j in temp:
                curr=T[j]
                curr.sort()
                for k in curr:
                    arr.append(k)
            Ans.append(arr)
        return Ans
                
            
            
                
            
                    

File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class comp{
   public: bool operator()(vector<int>&A,vector<int>&B)
    {
        if(A[1]<B[1])return true;
        if(A[1]==B[1])
        {
            if(A[0]<B[0])return true;
            if(A[0]==B[0])return A[2]<B[2];
            return false;
        }
        return false;
    }
};
class Solution {
    vector<vector<int>>A;
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        A.clear();
        traverse(root,0,0);
        sort(A.begin(),A.end(),comp());
        vector<vector<int>>ans;
        vector<int>temp;
        for(int i=0;i<A.size();i++)
        {
            if(i==0 or A[i-1][1]!=A[i][1])
            {
                if(temp.size()>0)ans.push_back(temp);
                temp.clear();
            }
            temp.push_back(A[i][2]);
        }
        if(temp.size()>0)ans.push_back(temp);
        return ans;
    }
private:
    void traverse(TreeNode *root,int i,int j)
    {
        if(root==nullptr)return ;
        else
        {
            A.push_back({i,j,root->val});
            traverse(root->left,i+1,j-1);
            traverse(root->right,i+1,j+1);
        }
    }
};